plugin/hooks.go

```diff
@@ -155,4 +156,8 @@ type Hooks interface {
 	// Note that this method will be called for files uploaded by plugins, including the plugin that uploaded the post.
 	// FileInfo.Size will be automatically set properly if you modify the file.
 	FileWillBeUploaded(c *Context, info *model.FileInfo, file io.Reader, output io.Writer) (*model.FileInfo, string)
+
+	// NotificationWillBeSent is invoked before a notification is sent.
+	// To modify the notified user map, return the replacement map,
+	NotificationWillBeSent(c *Context, post *model.Post, userList map[string]bool) map[string]bool
```

Comment:
I'm wondering if it makes more sense to move to a `PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification` interface instead. To me it's much cleaner what the hook does. For reference: This was proposed in https://github.com/mattermost/mattermost-server/pull/10344/files#diff-81d1bc59cb074b28cbda79d37dca9f56R166.

@rfoyard Would this also fulfill your use case? What do you think about this?

Reply:
@hanzei I think the concern I have with moving to that, is that you would have to call the plugin for every notification to every user. I thought that this would be a significant performance bottleneck.

I do agree that that interface would be more flexible. What do you think about the performance hit?

Reply:
I would like to see (in general) a hook that allows the editing of notifications. Given that I don't see a way around `PushNotificationWillBeSent(c *Context, notification *model.PushNotification) *model.PushNotification`.

Also, performance is not a huge deal to me. If notifications are delayed by e.g. 1 second, I don't mind.

I'm hesitant to make decision based on performance concerns until we measured what the "cost" of RPC is.

Reply:
@hanzei I think there's a misunderstanding here -- we're not talking about push notifications exactly, we're talking about the mentioned user list. Whether or not the user gets a push notification is up to their settings, their platform, etc. So hooking into that isn't what we're interested in here.

Reply:
Hm, with this in mind it makes sense. Can this hook also be used to notify a user that is not part of the channel?

Reply:
Yes it can, the goal of this hook is to both add and remove users to/from the list of mentioned users. As long as you know the userId, you can add it to the list.

Reply:
Nice. :+1: for the general approach then. Will do a full code review.